File: Trees/623 Add One Row to Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth==1:
            temp = TreeNode(val)
            temp.left=root
            return temp
        self.AddingTheRow(root,val,depth,1)
        return root
        
    
    def AddingTheRow(self,root,val,depth,level):
        
        if root==None:
            return
        
        if depth-1==level:
            
            temp = root.left
            root.left = TreeNode(val)
            root.left.left = temp
            temp = root.right
            root.right = TreeNode(val)
            root.right.right = temp
            
        else:
            
            self.AddingTheRow(root.left,val,depth,level+1)
            self.AddingTheRow(root.right,val,depth,level+1)
    

File: Trees/623 Add One Row to Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) 
    {
        if(depth==1)
        {
            TreeNode* temp = new TreeNode(val);
            temp->left=root;
            return temp;
        }
        helper(root,val,depth,1);
        return root;
    }
    
    void helper(TreeNode* root, int val,int depth, int maxdepth)
    {
        if(root==NULL)
            return;
            
        if(depth-1 == maxdepth)
        {
            TreeNode* temp = root->left;
            root->left = new TreeNode(val);
            root->left->left = temp;
            temp = root->right;
            root->right = new TreeNode(val);
            root->right->right = temp;
        }
        else
        {
            helper(root->left,val,depth,maxdepth+1);
            helper(root->right,val,depth,maxdepth+1);
        }
    }
};
